src/lib.rs

```diff
@@ -0,0 +1,269 @@
+use memtest::{MemtestError, MemtestOutcome, MemtestReport, MemtestType};
+use num_cpus;
+use rand::{seq::SliceRandom, thread_rng};
+use std::{
+    io::ErrorKind,
+    mem::size_of,
+    time::{Duration, Instant},
+};
+
+pub mod memtest;
+
+#[derive(Debug)]
+pub struct Memtester {
+    base_ptr: *mut usize,
+    memsize: usize,
+    timeout_ms: usize,
+    allow_working_set_resize: bool,
+    allow_mem_resize: bool,
+    allow_multithread: bool,
+    test_types: Vec<MemtestType>,
+}
+
+#[derive(Debug)]
+pub struct MemtestReportList {
+    pub tested_memsize: usize,
+    pub mlocked: bool,
+    pub reports: Vec<MemtestReport>,
+}
+
+#[derive(Debug)]
+pub enum MemtesterError {
+    WindowsWorkingSetFailure,
+}
+
+// TODO: get PAGE_SIZE from OS
+const PAGE_SIZE: usize = 4096;
+
+impl Memtester {
+    // TODO: Memtester without given base_ptr, ie. take care of memory allocation as well
+    // TODO: More configuration parameters:
+    //       early termination? terminate per test vs all test?
+
+    // NOTE: memsize may be decremented for mlock
+    // NOTE: memsize is advised to be a multple of size_of::<usize>() to avoid remainder bytes
+    /// Create a Memtester containing all test types in random order
+    pub fn new(
+        base_ptr: *mut u8,
+        memsize: usize,
+        timeout_ms: usize,
+        allow_mem_resize: bool,
+        allow_working_set_resize: bool,
+        allow_multithread: bool,
+    ) -> Memtester {
```

Comment:
I wonder if there is a better way to pass in configuration options than giving a bunch of bools. This current set up might make the code calling this function less readable?